lib/pos_leader_schedule.go

```diff
@@ -0,0 +1,106 @@
+	}
+
+	// Sum TotalStakeAmountNanos.
+	totalStakeAmountNanos := uint256.NewInt()
+	for _, validatorEntry := range validatorEntries {
+		totalStakeAmountNanos, err = SafeUint256().Add(totalStakeAmountNanos, validatorEntry.TotalStakeAmountNanos)
+		if err != nil {
+			return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error summing TotalStakeAmountNanos: ")
+		}
+	}
+
+	// Pseudocode for leader-selection algorithm:
+	// Note this is an O(N^2) algorithm where N is the number of validators we include.
+	// While len(LeaderSchedule) < len(ValidatorEntries):
+	//   Hash the CurrentRandomSeedHash and generate a new RandomUint256.
+	//   Take RandomUint256 modulo TotalStakeAmountNanos.
+	//   For each ValidatorEntry:
+	//     Skip if ValidatorEntry has already been added to the leader schedule.
+	//     If ValidatorEntry.TotalStakeAmountNanos >= RandomUint256:
+	//       Add ValidatorEntry to LeaderSchedule.
+	//       TotalStakeAmountNanos -= ValidatorEntry.TotalStakeAmountNanos.
+	//       Break out of the inner loop.
+	var leaderSchedule []*ValidatorEntry
+
+	// We also track a set of ValidatorPKIDs that have already been
+	// added to the LeaderSchedule so that we can skip them when
+	// iterating over ValidatorEntries in O(1) time.
+	leaderSchedulePKIDs := NewSet([]*PKID{})
+
+	for len(leaderSchedule) < len(validatorEntries) {
+		// Hash the CurrentRandomSeedHash each iteration. This generates
+		// multiple predictable pseudorandom values from the same seed.
+		currentRandomSHA256 := sha256.Sum256(currentRandomSeedHash.ToBytes())
+		currentRandomSeedHash, err = (&RandomSeedHash{}).FromBytes(currentRandomSHA256[:])
+		if err != nil {
+			return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error hashing CurrentRandomSeedHash: ")
+		}
+
+		// Take RandomUint256 % TotalStakeAmountNanos.
+		randomUint256 := uint256.NewInt().Mod(currentRandomSeedHash.ToUint256(), totalStakeAmountNanos)
+
+		// Keep track of the stake seen so far in this loop.
+		sumStakeAmountNanos := uint256.NewInt()
+
+		for _, validatorEntry := range validatorEntries {
+			// Skip if ValidatorEntry has already been added to the leader schedule.
+			if leaderSchedulePKIDs.Includes(validatorEntry.ValidatorPKID) {
+				continue
+			}
+
+			// Sum the ValidatorEntry.TotalStakeAmountNanos to the stake seen so far.
+			sumStakeAmountNanos, err = SafeUint256().Add(sumStakeAmountNanos, validatorEntry.TotalStakeAmountNanos)
+			if err != nil {
+				return nil, errors.Wrapf(err, "UtxoView.GenerateLeaderSchedule: error summing TotalStakeAmountNanos: ")
+			}
+
+			// If the sum of the stake seen so far is less than the RandomUint256, skip this validator.
+			if sumStakeAmountNanos.Lt(randomUint256) {
+				continue
+			}
+
+			// If we get to this point, the current validator is the
+			// one we should add to the leader schedule next.
+
+			// Add the current ValidatorEntry to the leaderSchedule.
+			leaderSchedule = append(leaderSchedule, validatorEntry)
```

Comment:
could we just remove the validatorEntry from the slice of validatorEntries once its been added to the leader schedule? we could get rid of the set with this change.

Reply:
That's the way I initially had it like `validatorEntries = append(validatorEntries[:index], validatorEntries[index+1:])`, but after reading into it, IMO the edge case handling around slice indexing is tricky and the space/time complexity of doing that slice concatenation is no better than using the set IMO. LMK what you think.

Reply:
Put another way, we have a trade-off here between space vs. time. We can triplicate the list of ValidatorEntries: 1) the input validator entries, 2) the leader schedule slice, and 3) the set of validator entries already added to the leader schedule. But look-ups into the set of already added validators is O(1). Alternatively, we can save ourselves from also keeping that set in memory and instead remove the validator from the input slice of validator entries. This saves space but costs us some time/runtime since the algorithm to remove an element from a slice is not O(1). IMO since the number of validators we consider in our leader schedule generation algorithm is capped (currently to 100), it's worth taking up more space to save on some time. Note that the algorithm is already O(n^2).